### PR TITLE
DEV-4957-code-keeper

### PR DESCRIPTION
**Pull request by zamazzal**
> Before starting, please choose the relevant pull request **Labels**, **Reviewers**, and **Assignees**

### Why?
add devops project for devops branch
> Clearly state the reason for this change. What problem is it solving or feature is it adding?

### Solution Overview
add a project for devops to leet the learners create a full devops pipelines

> Provide an overview of the solution implemented in this pull request. This should be a high level overview without getting into technical details. If applicable, include screenshots of UI or use GitHub compliant [`mermaid`](https://mermaid.live/) graphs to visually represent the solution.

### Implementation Details

> Explain the details of the implementation and the reasoning behind it. What alternative approaches were considered and why was this approach chosen?

### CODE REVIEW

File: subjects/devops/code-keeper/README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
+## Code-Keeper
+
+<center>
+<img src="./resources/cloud-design.jpg?raw=true" style = "width: 400px !important; height: 400px !important;"/>
+</center>
+
+### Objective
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:7]: This section is to explain the objectives of the project, I add tasks in order to understand the objectives.

File: subjects/devops/code-keeper/README.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+<center>
+<img src="./resources/cloud-design.jpg?raw=true" style = "width: 400px !important; height: 400px !important;"/>
+</center>
+
+### Objective
+
+In this project, you will create a complete pipeline to scan and deploy a microservices-based application using Node.js. Your challenge is to design, implement, and optimize a pipeline that incorporates industry best practices for continuous integration, continuous deployment, and security. Your mission is to:
+
+- Set up a source control system for the Node.js microservices source code and the infrastructure configuration.
+- Create a Pipeline to create, update, or delete the infrastructure for the staging and production environment.
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:12]: fixed

File: subjects/devops/code-keeper/README.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+- Create a continuous integration (CI) pipeline to build, test, and scan the source code.
+- Create a continuous deployment (CD) pipeline to deploy the application to a staging and production environment.
+- Ensure the security and reliability of the application throughout the pipeline stages.
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:15]: fixed

File: subjects/devops/code-keeper/README.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+To complete this project, you should have a good understanding of the following:
+
+- Node.js and its ecosystem
+- Gitlab and Gitlab CI
+- Ansible as a configuration management and automation tool
+- Docker and containerization
+- Terraform as an Infrastructure as Code (IaC)
+- Cloud platforms (e.g., AWS, Azure, or Google Cloud)
+
+### Tips
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:28]: in this case, it's tips not hints

File: subjects/devops/code-keeper/README.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+The goal of the roleplay question session is to:
+
+- Assess your understanding of the concepts and technologies used in the project.
+- Test your ability to communicate effectively and explain your decisions.
+- Challenge you to think critically about your solution and consider alternative approaches.
+
+Prepare for a roleplay question session where you will assume the role of a DevOps engineer presenting your solution to your team or a stakeholder. You should be ready to answer questions and provide explanations about your decisions, architecture, and implementation.
+
+### Deploy GitLab and Runners for Pipeline Execution
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:49]: done

File: subjects/devops/code-keeper/README.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+- `Plan`: Generate an execution plan that shows the changes to be made to your infrastructure, including the resources that will be created, updated, or deleted. This job provides a preview of the changes and enables you to review them before applying.
+
+- `Apply to Stagging`: Apply the Terraform configuration to create, update, or delete the resources as specified in the execution plan. This job provisions and modifies the infrastructure in the staging environment.
+
+- `Approval`: Require manual approval to proceed with deployment to the `production environment`. This step should involve stakeholders and ensure the application is ready for production.
+
+- `Apply to Production`: Apply the Terraform configuration to create, update, or delete the resources as specified in the execution plan. This job provisions and modifies the infrastructure in the production environment.
+
+3. Design and implement a CI pipeline for each repository that will be triggered on every code push or pull request. The pipeline should include the following stages:
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:95]: fixed

File: subjects/devops/code-keeper/README.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+- `Deploy to Staging`: Deploy the application to a `staging environment` for further testing and validation.
+- `Approval`: Require manual approval to proceed with deployment to the `production environment`. This step should involve stakeholders and ensure the application is ready for production.
+- `Deploy to Production`: Deploy the application to the `production environment`, ensuring zero downtime and a smooth rollout.
+
+> Each repository must have a pipeline!
+
+> Any modification in the application's source code must rebuild and redeploy the new version to the `Staging Environment` and then to the `Production Environment` after manual approval.
+
+### Cyber Security
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:112]: done

File: subjects/devops/code-keeper/README.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+- `Apply to Production`: Apply the Terraform configuration to create, update, or delete the resources as specified in the execution plan. This job provisions and modifies the infrastructure in the production environment.
+
+3. Design and implement a CI pipeline for each repository that will be triggered on every code push or pull request. The pipeline should include the following stages:
+
+- `Build`: Compile and package the application.
+- `Test`: Run unit and integration tests to ensure code quality and functionality.
+- `Scan`: Analyze the source code and dependencies for security vulnerabilities and coding issues. Consider using tools such as `SonarQube`, `Snyk`, or `WhiteSource`.
+- `Containerization`: Package the applications into Docker images using a Dockerfile, and push the images to a container registry (e.g., Docker Hub, Google Container Registry, or AWS ECR).
+
+4. Design and implement a CD pipeline that will be triggered after the CI pipeline has been completed. The pipeline should include the following stages:
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:102]: done

File: subjects/devops/code-keeper/audit/README.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+##### Review the Documentation:
+
+###### Does the `README.md file contain all the necessary information about the solution (prerequisites, setup, configuration, usage, ...)?
+
+###### Is the documentation provided by the student clear and complete, including well-structured diagrams and thorough descriptions?
+
+#### Bonus
+
+###### +Did the student add any optional bonus?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:122]: fixed

File: subjects/devops/code-keeper/audit/README.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+###### Have the students separated credentials from code, using secure methods like secret management tools or environment variables?
+
+###### Did the students apply the least privilege principle to limit user and service access to the minimum required level?
+
+###### Do the students have a process for updating dependencies and tools regularly, automating updates, and monitoring for security advisories and patches?
+
+##### Review the Documentation:
+
+###### Does the `README.md file contain all the necessary information about the solution (prerequisites, setup, configuration, usage, ...)?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:116]: fixed

File: subjects/devops/code-keeper/audit/README.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+##### Review the GitLab and Runners Deployment:
+
+###### Was the GitLab instance deployed and configured successfully using Ansible?
+
+###### Are the GitLab Runners integrated with the existing pipeline and executing tasks as expected for all repositories?
+
+##### Review the Infrastructure Pipeline:
+
+###### Does the student deploy the `cloud-design` project infrastructure for two environments (staging, prod) on a cloud platform (e.g., AWS, Azure, or Google Cloud) using `Terraform`?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:65]: done

File: subjects/devops/code-keeper/audit/README.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+###### Do the two environments are similar in design, resources, and services used?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:67]: done

File: subjects/devops/code-keeper/audit/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:65]: done

File: subjects/devops/code-keeper/audit/README.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+###### Are the Build, Test, Scan, and Containerization stages implemented correctly in the CI pipeline for each repository?
+
+##### Review the CD Pipeline:
+
+- `Deploy to Staging`: Deploy the application to a `staging environment` for further testing and validation.
+- `Approval`: Require manual approval to proceed with deployment to the `production environment`. This step should involve stakeholders and ensure the application is ready for production.
+- `Deploy to Production`: Deploy the application to the `production environment`, ensuring zero downtime and a smooth rollout.
+
+###### Are the Deploy to Staging, Approval, and Deploy to Production stages implemented correctly in the CD pipeline for each repository?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:90]: done

File: subjects/devops/code-keeper/audit/README.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+##### Review the Infrastructure Pipeline:
+
+###### Does the student deploy the `cloud-design` project infrastructure for two environments (staging, prod) on a cloud platform (e.g., AWS, Azure, or Google Cloud) using `Terraform`?
+
+###### Do the two environments are similar in design, resources, and services used?
+
+###### Does the student's infrastructure configuration exist in an independent repository with a configured pipeline?
+
+###### Are the Init, Validate, Plan, Apply to Staging, Approval, and Apply to production environment stages implemented correctly in the infrastructure pipeline?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:71]: done

File: subjects/devops/code-keeper/audit/README.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+###### Are the Init, Validate, Plan, Apply to Staging, Approval, and Apply to stagging environment stages implemented correctly in the infrastructure pipeline?
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:73]: done

File: subjects/devops/code-keeper/audit/README.md [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+###### Are the students capable of thinking critically about their solution and considering alternative approaches?
+
+##### Review the GitLab and Runners Deployment:
+
+###### Was the GitLab instance deployed and configured successfully using Ansible?
+
+###### Are the GitLab Runners integrated with the existing pipeline and executing tasks as expected for all repositories?
+
+##### Review the Infrastructure Pipeline:
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/audit/README.md:63]: not necessary questions to check the applications is already done in the other projects audit.

File: subjects/devops/code-keeper/README.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+
+You must deploy a GitLab instance using Ansible. This hands-on exercise will help you gain a deeper understanding of Ansible as a configuration management and automation tool while also giving you experience in deploying and configuring GitLab.
+
+1. Create an Ansible playbook to deploy and configure a GitLab instance. The playbook should automate the installation of GitLab and any required dependencies. It should also configure GitLab settings such as user authentication, project settings, and CI/CD settings.
+
+2. Deploy a GitLab instance on a cloud platform (e.g., AWS, Azure, or Google Cloud) or in a local environment using the Ansible playbook. Ensure that the instance is accessible to all team members and is configured to support collaboration and code reviews.
+
+3. Configure the GitLab instance to support CI/CD pipelines by setting up GitLab Runners and integrating them with your existing pipeline. Update your pipeline configuration to utilize GitLab CI/CD features and execute tasks on the deployed Runners.
+
+4. Demonstrate the successful deployment and configuration of GitLab using Ansible.
```
[the remark this answers is not in the document]
Comment (zamazzal) [reply subjects/devops/code-keeper/README.md:59]: done

**Review by MSilva95 — APPROVED** [review]
LGTM!

**Pull request merged**